### ISSUE TITLE
Typescript support (or at least declarations) inside Cinnamon

### ISSUE THREAD

**Issue by Gr3q** [linuxmint/cinnamon#11272]
### Is your feature request related to a problem? Please describe.

This includes allowing writing code in Typescript not just in Javascript. This includes:
1. Provide Typescript declarations under `imports.gi`, generating Cinnamon specific libs during build and provide for those we explicitly depend on. 
2. Provide Typescript declarations in the JS folder as `d.ts`
3. Allow writing code in typescript and generating `js` and `d.ts` files during build - this might need some extra effort so typescript knows about those libraries in `imports`.

### Describe the solution you'd like

1. I've got a [repository](https://github.com/Gr3q/GIR2TS) that does just that (output is stored [here](https://github.com/Gr3q/types-cjs)), it is pretty solid and used with some applets already. I'm happy to transfer over the repo to the Cinnamon team and still maintaining it myself.
2. This is a relatively easy task if the 1st step is done (technically 1. can be provided outside of Cinnamon itself for development) - downside is that these will be maintained alongside the JS files.  I'm more than happy to commit to write them and maintain them myself.
3. If imports are correctly resolve before (in TS) and after build (in JS) the codebase can be slowly converted .

### Additional context - why?

1. Personally I'd like to add more features to `js->ui->windowManager`, this is why I opened this request.
2. Could reduce regressions in the JS codebase because it will be typed.
4. Could provide better support for applets/extensions/desklets with correct types in editors with docstrings
5. Could reduce cognitive load to people working on the JS codebase and maybe help people open more PRs because the codebase functionality is easier to read, because everything in `imports` is not opaque as it is now or requires to read up documentation (if it exists - I'm looking at the Cinnamon specific libs in imports especially.)
6. Could enforce better code quality

### Possible problems
1. One more thing added to the toolchain Cinnamon is built with
2. (Slightly) higher level of entry of writing JS code and a bit more work to setup the dev environment 

Please respond if you think it's worth having a discussion about, especially that I can pledge to to most of the work around this.

**Comment by ItzSwirlz** [comment]
This would be interesting though I'm not sure what usage this would cover other than being ideal. GJS's documentation mentions how it deals with C types

**Comment by Gr3q** [comment]
`ts-for-gir` is for `gjs` or `node`, it generates declarations differently so it's not compatible with `cjs`, but maybe I'm wrong.

Also it doesn't generate docstrings.

**Comment by NathanC** [comment]
This would be amazing. I'm interested in contributing, and having proper TypeScript support would make me far more likely to contribute. I think it would be great for more modern development of applets/Desklets/extensions, and may grow the community.

It also would rely in more reliable code, as TypeScript helps eliminate a large class of runtime bugs.

**Comment by mtwebster** [comment]
This seems great for any pure JS api that we have (applet definitions, etc...), but I really don't understand why this is needed for our introspected libraries.

We already have documentation for this stuff. It's all written in C so it's clear what argument and return types are, and the C code will generate warnings if we do things like send a string when it wants an int. There are resources to help those less familiar with it figure out how `st_icon_set_icon_name` becomes St.Icon.set_icon_name().

This seems like reinventing the wheel to me. A better job could definitely be done explaining how introspection works, and tying everything together, but I don't see why we need all of this - there are already 'guard rails' in place here. I don't want to add things just 'because'. What am I missing?

disclaimer: I know JS only from working on Cinnamon and a couple of other minor, non-web-related applications.

https://gjs-docs.gnome.org/
https://lazka.github.io/pgi-docs/

 
 > ts-for-gir is for gjs or node, it generates declarations differently so it's not compatible with cjs, but maybe I'm wrong.

cjs is gjs with some renaming. We made 'cjs' to we could avoid being forced into new versions of gjs before we were ready (particularly important for us when supporting both Ubuntu and Debian-based Mint versions) - back then gjs was evolving a lot from version to version.

**Comment by Gr3q** [comment]
>This seems great for any pure JS api that we have (applet definitions, etc...), but I really don't understand why this is needed for our introspected libraries.

I can give you an itemized list of why it's important and also on the other points:
* For typescript to work it needs to know the types of things. If I mark `imports.gi` as `any` because we don't have types, I disable Typescript type checking completely. And that spreads into the pure JS codebase because it relies so heavily on the introspection libraries
* https://gjs-docs.gnome.org/ is excellent documentation for _gjs_ (https://lazka.github.io/pgi-docs/ for cjs as well in some level). Where is the documentation on `Cinnamon`? Where is the Cinnamon version -> introspection lib version mapping? I'm left with reading the source code (or compile cinnamon for myself where I generate the `.gir` files myself and browse that).
* Typescript will compile TS to JS and issue errors and warning based on the types are given and can be configured to be a lot stricter than JS. This means you can catch errors **build time**, not runtime. This means:
   * [this](https://github.com/linuxmint/cinnamon/pull/11131) would have never happened, Typescript would have caught that the moment you regenerated the declarations for the new lib versions used in 5.4 and tried to compile ts files again.
   * In my POC [here](https://github.com/linuxmint/cinnamon/pull/11276/files#diff-600d50677fc4736c5644f03946d807c8c3cbc8ca4b0d13eb31fdefcc67b6ffd2R134) I've already found that `GLib.shell_parse_argv` can return null for `argv`, and `trySpawn` passed that straight into `GLib.spawn_async` that doesn't accept null, and that will either throw (or issue a warning as your said).
   * You can extrapolate this to the pure JS codebase

The important takeaway is that if everything is typed, Typescript will prevent errors **build time** instead of finding errors later runtime (or if it's something seldom used in the JS codebase, in production).

>cjs is gjs with some renaming. We made 'cjs' to we could avoid being forced into new versions of gjs before we were ready (particularly important for us when supporting both Ubuntu and Debian-based Mint versions) - back then gjs was evolving a lot from version to version.

I'll keep that in mind, in that case it might be worthwhile for me to repurpose https://gitlab.gnome.org/JumpLink/gi.ts instead of writing [GIR2TS](https://github.com/Gr3q/GIR2TS) myself. `gi.ts` doesn't have docstrings and seem to inject `$gtype` into every object's toplevel it generated (idk if that's correct, in cinnamon's codebase I see it's accessed from the `constructor`), on the other hand, my tool has a terrible structure (codebase-wise)...

**Comment by mtwebster** [comment]
Could the .ts files for external libraries be generated at build time? The gir files come with the -dev packages typically, and I'm really not crazy about storing those from other libraries.

Cinnamon is used on many different distros (usually with many different versions of these libraries as well as the introspection 'plumbing' itself, so having them here statically could cause problems I think.

Maybe some sort of combination of the two - keep the generated files in the repo, but also make sure they match the library versions used at build-time, and regen if needed. Maybe the compile step already does some sort of checking? Sort of irrelevant here, but this sounds like something that would best be created when g-ir-scanner is run on the library builds themselves.

> (Slightly) higher level of entry of writing JS code and a bit more work to setup the dev environment

Some sort of 'cinnamon-compile-js' script for xlet devs? Or something more?

I'm due to bring switch to a newer mozjs/cjs, are there changes in more recent gjs versions that might simplify any of this?

Thanks

**Comment by Gr3q** [comment]
>TL;DR: generating TS for this would essentially necessitate generating TS for all of Gtk, and this repo is not a place to do that. There needs to be a structured area for all Gtk/GLib TS gir generation, then we can do this

Agreed

>Even if was generated at build time, it would take a while, and require more dependencies, and another script to execute which may not produce reliable results. Or, it just is not worth fitting in with default and having support for everyone who installs it, especially if it is not necessary. (See the python xapp bindings) But even if you ignored that, it would dramatically change the structure of this repo, which is already pretty huge.

* My current script takes about 5 seconds to generate the types for all that you see in the PR and it doesn't actually need the dependencies to generate, because it automatically generates to the `imports` namespace (up to you to decide if it's actually a good thing).
* I agree, anything that doesn't have the source code in this repo should not be stored here
* I think Cinnamon, CinnamonDesktop and basically anything that we generate `.gir` files from here should also generate the typescript declarations (but not actually stored! just like the `.gir` files) because they actually belong to this repo.

Compiling from TS will JS (build time, PR validation etc) will be a part of the build process and that needs the declarations to be in place when it happens (at that point they either need to be included from a specific place and/or generated).

In the end I agree that most of them need to be stored/included from somewhere else because same as GLib and the others, there are dependencies all the same. If you can dynamically get the `.gir` files it's better, if not they need to be stored somewhere. 

@mtwebster 
>I'm due to bring switch to a newer mozjs/cjs, are there changes in more recent gjs versions that might simplify any of this?

Most likely not

>Cinnamon is used on many different distros (usually with many different versions of these libraries as well as the introspection 'plumbing' itself, so having them here statically could cause problems I think.

I though Cinnamon uses specific versions of them that are known.

**Comment by mtwebster** [comment]
We only use specific versions of libraries that we publish - libcinnamon-menus, -desktop, cvc (gvc), and those in cinnamon and muffin - clutter, cogl, muffin, st. Everything else is system.

**Comment by Gr3q** [comment]
I'm trying to compile the 2 lists now

So these are provided by Cinnamon:
* CDesktopEnums-3.0
* Cinnamon-0.1
* CinnamonDesktop-3.0
* Clutter-7
* Clutter-X11-7 - is this provided by Cinnamon as well
* CMenu-3.0
* Cogl-7
* CoglPango-7
* Cvc-1.0
* St-1.0
* Muffin-? - I haven't got the .gir file for this actually

Not provided by Cinnamon but depends on:
* Atk-1.0
* cairo-1.0
* Gdk-3.0
* GdkPixbuf-2.0
* Gio-2.0
* GIRepository-2.0
* GLib-2.0
* GL-1.0
* GModule-2.0
* GObject-2.0
* Graphene-1.0
* Gtk-3.0
* HarfBuzz-0
* Meta-7
* Pango-1.0
* PangoCairo-1.0
* Soup-2.4
* Json-1.0
* xlib-2.0
* xfixes-4.0

Let me know if I missed something or made a mistake.

btw There are some errors in some of Cinnamon's generated .gir files, for example Cinnamon provides incorrect versions for it's dependencies:

```xml
  <include name="Clutter" version="0"/>
  <include name="ClutterX11" version="0"/>
  <include name="Cogl" version="0"/>
  <include name="CoglPango" version="0"/>
  <include name="Meta" version="0"/>
```

and it reports that it's using `Clutter.AllocationFlags` in `GenericContainer` `allocate` signal's callback parameters while Clutter doesn't have that enum anymore.

**Comment by Gr3q** [comment]
What do I need to do to progress this? 

I'll just write down something that resembles a "proposal" to get some feedback.

## Packaging

My idea would be to generate a package named something like `gir-1.0-ts-cinnamon-typestubs` and provide declarations in `/usr/share/gir-1.0-ts-cinnamon` folder. Naming to be decided. This is fine because we don't depend on specific versions but major versions of libraries. If the typestubs `.gir` filelist are updated (it can be just a filename list then they are sourced from `/usr/share/gir-1.0`) when cinnamon changes it's dependencies then it should be fine - you know, `Soup-2.4`->`Soup-3`.

In there declarations can be generated and packaged, so they can be installed as a make dependency or independently by devs.

Cinnamon and other packages can generate their own declarations and can be placed in the same folder - `/usr/share/gir-1.0-ts-cinnamon` - during install.

I named the folder `/usr/share/gir-1.0-ts-cinnamon` so it's known that it's provided specifically for Cinnamon and only contains the minimum libs required/provided by Cinnamon.

## Building packages that build depends on `gir-1.0-ts-cinnamon-typestubs`
The only thing to work out is to develop a build steps for packages that depend on `gir-1.0-ts-cinnamon-typestubs`.

### Requirements
* `npm`, `nodejs` and `gir-1.0-ts-cinnamon-typestubs` as build dependencues
* A `tsconfig.json` file that includes (in build) the following folders:
   * `/usr/share/gir-1.0-ts-cinnamon`
   * a folder where all generated `d.ts` files are placed during `.gir`->`.d.ts` build for said package - can be different for each package if needed.

### Steps

The steps I came up with (for cinnamon as an example)
...
1. Install new dependencies
2. Build everything that generates .gir files
3. Generate `d.ts` files from .gir files and place them into the correct folder
4. Typescript build
...
5. Install generated `.d.ts` files into `/usr/share/gir-1.0-ts-cinnamon`

## Building packages that only provide `.d.ts` files

These packages don't contain Typescript files, only provide `d.ts` files.

### Requirements
* `npm` and `nodejs` as build dependencies - the only reason because I've written the generator tool in Typescript, if I would rewrite it in another language it could be removed (don't ask me to rewrite it in C or C++, the best I can do for you is Rust or Python)

### Steps
...
1. Install new dependencies
2. Build everything that generates .gir files
3. Generate `d.ts` files from .gir files and place them into the correct folder
...
4. Install generated `.d.ts` files into `/usr/share/gir-1.0-ts-cinnamon`

**Comment by Gr3q** [comment]
Modifications to the proposal above and my current recommendation:

* Fork `gi.ts` and only add the minimum necessary fixes needed (which I can do)
* Correctly provide all `.gir` files from all packages (I think only `cinnamon` doesn't do that, https://github.com/linuxmint/cinnamon/pull/11335)
* Only use the forked `gi.ts` in packages hat actually has `js`/`ts` code. Currently that means `cinnamon` only (I think?).  The steps for this:
    1. Dependencies for `cinnamon` already build and providing the necessary `.gir` files on the system.
    2. Have `d.ts` files in place in `cinnamon` that describes the changes made to `cjs` global scope - `imports`, `global` consts and so on
    3. _Build step_: `cinnamon` already build it's own `.gir` files as well
    4. _Build step_: Generate direct library dependencies as `d.ts` files with the forked `gi.ts` tool as the next step in build - configured to search `.gir` files from `cinnamon` itself, in `/usr/share/gir-1.0` and `/usr/lib/muffin`.
    6. Build `.ts`->`.js` build in `cinnamon` using the built declarations.
* Extra: Provide the built `.d.ts` files with `cinnamon` at a standard location applet/desklet/extension devs can use - obviously this only ensures that the applets work on their system, but it's a good start.

**Comment by raluvy95** [comment]
Any progress on this? Would be really great to use TypeScript for Cinnamon applet/desklet/extension developement.

**Comment by Gr3q** [comment]
@raluvy95 This issue is about using/integrating typescript into Cinnamon itself. While it would be nice, reduce the maintenance burden of separately maintaining declaration files, it's not necessary for extension development.

Why there is no interest for integration:
* I guess it's not deemed that important
* I was using my own tool to generate types and something more established is desired - https://github.com/gjsify/ts-for-gir is something like it but last time I checked it was missing features/calls my spaghetti tool had.

Still you could use https://github.com/Gr3q/types-cjs to develop Typescript extensions. The 2 examples I know of https://github.com/linuxmint/cinnamon-spices-applets/tree/master/radio%40driglu4it https://github.com/linuxmint/cinnamon-spices-applets/tree/master/weather%40mockturtl that are using it.

**Comment by sharno** [comment]
I'd really love to see cinnamon have TS
